x-pack/legacy/plugins/uptime/server/lib/adapters/monitor_states/get_snapshot_helper.ts

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { MonitorGroups, MonitorGroupIterator } from './search';
+
+export const getSnapshotCountHelper = async (iterator: MonitorGroupIterator) => {
+  const items: MonitorGroups[] = [];
+  let res: MonitorGroups | null;
+  // query the index to find the most recent check group for each monitor/location
+  do {
+    res = await iterator.next();
+    if (res) {
+      items.push(res);
+    }
+  } while (res !== null);
+  return (
+    items
+      // result is a list of 'up' | 'down', 1:1 per monitor
+      .map(({ groups }) =>
+        // for each location, infer a status
+        groups.reduce<'up' | 'down'>((acc, cur) => {
+          if (acc === 'down') {
+            return acc;
+          }
+          return cur.status === 'down' ? 'down' : 'up';
+        }, 'up')
+      )
+      // count each status up into a single object
+      .reduce(
+        (acc, cur) => {
```

Comment:
While I do appreciate functional code, I do think it would be more readable to use imperative code here. It would be

```javascript
let up = 0;
let down = 0;
items.forEach( (item) => {
  item.groups.find(g => g.status === 'down') ? down++ : up ++
})
return {up, down}
```

The pattern of using reduce for counters is often more awkward I've come to find. It can still make sense in situations with an unknown number of keys, but IMHO imperative is easier to read here.